modules/n1ql/pages/n1ql-language-reference/recursive.adoc

```diff
@@ -0,0 +1,315 @@
+options:: Object that signals an early exit.
+* `levels`: An exit after level `N`. 
+This object assumes the Anchor is at level 0.
+* `documents`: An exit after accumulating `N` documents. 
+
+[source,ebnf]
+----
+include::partial$grammar/dql.ebnf[tag=recursive-with-clause-cycle]
+----
+
+image::n1ql-language-reference/recursive-with-clause-cycle.png["Argument diagram", align=left]
+
+cycle:: `CYCLE <fieldnames_list> RESTRICT`. 
+List of field names in which you want to perform cycle detection.
+
+[#limitations]
+== Limitations
+
+* The recursive reference is only allowed once in the FROM clause.
+It's not allowed anywhere else.
+
+* ORDER BY, LIMIT, and OFFSET clauses are not allowed in the SELECT statement in the subquery used to define the anchor and the recursive clause.
+
+* The DISTINCT quantifier is not allowed in anchor and recursive clauses.
+
+* GROUP BY, WINDOW, and AGGREGATE functions are not allowed in recursive clauses.
+
+* OUTER JOINS are not allowed in recursive clauses because they can lead to potential infinite recursion.
+
+* Recursive clauses do not support NEST and UNNEST clauses.
+
+* If there is no UNION/UNION ALL separation for the recursive CTE, the query defaults to a normal CTE.
+
+* A syntax error is returned when optional subclauses are used without the RECURSIVE keyword.
+
+* In general, recursion is also limited by:
+** The logic in the recursive statement.
+** The stop in the options argument.
+** Breaching the request timeout, if configured.
+** Breaching the request memory quota, if configured.
+** Exceeding the implicit document limit (10000) when a memory quota is not in use and when no explicit document limit is set in the options.
+** Exceeding the implicit level limit (1000) when the level option is not in use.
+
+[#examples_section]
+== Examples
+
+The following examples follow linear recursion.
+Only one recursive reference is allowed in the FROM clause.
+Self-joins or set-ops are not allowed in the recursive reference.
+
+.Example of simple recursive referencing
+====
+[source,sqlpp]
+----
+WITH RECURSIVE cte AS (
+    SELECT 1 AS r 
+        UNION 
+    SELECT cte.r+1 AS r 
+    FROM cte 
+    WHERE cte.r<4 
+) 
+SELECT cte.r FROM cte;
+----
+
+.Results
+[source,json]
+----
+[
+  {
+    "r": 1
+  },
+  {
+    "r": 2
+  },
+  {
+    "r": 3
+  },
+  {
+    "r": 4
+  }
+]
+----
+====
+
+.Example combining recursive and non-recursive CTEs with the WITH clause
+====
+[source,sqlpp]
+----
+WITH RECURSIVE cte AS (SELECT 1 r) ,
+    rcte AS ( 
+        SELECT cte.r FROM cte 
+            UNION 
+        SELECT rcte.r+2 r FROM rcte WHERE rcte.r<7
+    )
+SELECT * FROM rcte;
+----
+
+.Results
+[source,json]
+----
+[
+  {
+    "rcte": {
+      "r": 1
+    }
+  },
+  {
+    "rcte": {
+      "r": 3
+    }
+  },
+  {
+    "rcte": {
+      "r": 5
+    }
+  },
+  {
+    "rcte": {
+      "r": 7
+    }
+  }
+]
+----
+====
+
+.Example finding all possible routes leaving from the TLV airport and going to 2 other airports
```

Comment:
Add this before, as recursive queries tend to take long without appropriate indexes.
CREATE INDEX srcdestidx ON `travel-sample`.inventory.route( sourceairport, destinationairport );

Reply:
Hi,
I can confirm that an index -> def_inventory_route_route_src_dst_day :
Definition  ```CREATE INDEX `def_inventory_route_route_src_dst_day` ON `travel-sample`.`inventory`.`route`(`sourceairport`,`destinationairport`,(distinct (array (`v`.`day`) for `v` in `schedule` end))) WITH { "defer_build":true }```

This index is used by the query but the runtime is roughly 14seconds on my machine

But the above index-> ```CREATE INDEX srcdestidx ON travel-sample.inventory.route( sourceairport, destinationairport );```

Results in a faster query( 5seconds something). I would prefer that this index be created:)